### PR TITLE
[WIP][Feature] Server excluded bad disk

### PR DESCRIPTION
**Pull request by jerqi**

### CODE REVIEW

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java [hunk 1]
```diff
@@ -145,6 +150,7 @@ public void addToFlushQueue(ShuffleDataFlushEvent event) {
 
   private void flushToFile(ShuffleDataFlushEvent event) {
 
+    event.pickStorageIndex(storageBasePaths.length);
```
Comment (colinmjj) [inline server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:153]: If storage is broken during the shuffle, shuffle data will be located in different location for the same partition?

Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:153]: Yes, partition will locate an another storage. Actually, once a storage is broken,  I hope shuffle will fail fast. I hope to implement this function.

**Comment by duanmeng** [comment]
Could you write a description about your strategy for juding whether a disk is damaged or repaired and the following actions such as the recovery procedure. And what is this pr's goal.

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java [hunk 2]
```diff
@@ -156,86 +162,98 @@ private void flushToFile(ShuffleDataFlushEvent event) {
...
+  }
+
+  private void processWriteResult(ShuffleDataFlushEvent event, long duration, boolean writeSuccess) {
+    if (shuffleServer != null) {
+      shuffleServer.getShuffleBufferManager().releaseMemory(event.getSize(), true, false);
+      storageErrorMap.putIfAbsent(event.getStorageIndex(), new AtomicInteger(0));
+      AtomicInteger errors = storageErrorMap.get(event.getStorageIndex());
+      if (writeSuccess) {
+        errors.updateAndGet(x -> {
+          --x;
```
Comment (duanmeng) [inline server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:212]: `errors.updateAndGet(x -> --x < 0 ? 0 : x)`?

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java [hunk 3]
```diff
@@ -156,86 +162,98 @@ private void flushToFile(ShuffleDataFlushEvent event) {
...
+          if (x < 0) {
+            return 0;
+          }
+          return x;
+        });
+        LOG.debug("Flush to file success in " + duration + " ms and release " + event.getSize() + " bytes");
```
Comment (duanmeng) [inline server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:218]: use log placeholder

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java [hunk 4]
```diff
@@ -156,86 +162,98 @@ private void flushToFile(ShuffleDataFlushEvent event) {
...
+      } else {
+        if (storageBasePaths.length > 1 && errors.incrementAndGet() > excludeStorageThreshold) {
+          ShuffleStorageUtils.addExcludeStorage(event.getStorageIndex());
+          LOG.error("Add {} to exclude storage set", storageBasePaths[event.getStorageIndex()]);
         }
+        LOG.error("Flush to file for " + event + " failed in "
```
Comment (duanmeng) [inline server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:224]: use log placeholder

**Comment by jerqi** [comment]
> Could you write a description about your strategy for juding whether a disk is damaged or repaired and the following actions such as the recovery procedure. And what is this pr's goal.

No repairing operation. I will update pr's description according to template.

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java (hunk 4 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:224]: OK

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java (hunk 3 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:218]: OK

File: server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java (hunk 2 above)
Comment (jerqi) [reply server/src/main/java/com/tencent/rss/server/ShuffleFlushManager.java:212]: I prefer the current implement. Because it's more easy to understand.

**Comment by jerqi** [comment]
In our production environment, we find  that raid disks can provide the fault tolerence and better performance. So the pr seems not to be much important.

**Pull request closed without merging**